### PR TITLE
[Data] reddit data loader

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description
TODO: upload .npz files to S3.
Currently we need manually generate and copy `reddit_graph.npz` and `reddit_data.npz` to `~/.dgl/`.

Then we can use it as `python3 load_reddit.py --dataset reddit`

```
### load_reddit.py

import argparse
import numpy as np
import time
import mxnet as mx
from mxnet import gluon
import dgl
from dgl.data import register_data_args, load_data


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='GCN')
    register_data_args(parser)
    args = parser.parse_args()

    print(args)

    data = load_data(args)
    in_degree = [data.graph.in_degree(i) for i in range(len(data.graph))]
    print(data.features.shape, data.labels.shape)
    print(data.graph.number_of_nodes(), data.graph.number_of_edges())
```

Script to get `reddit_graph.npz` and `reddit_data.npz`:
```
wget http://snap.stanford.edu/graphsage/reddit.zip
unzip reddit.zip
cd reddit
# pip3 install --user networkx==1.11
python3 convert_to_npz.py
```

```
### convert_to_npz.py

import numpy as np
import scipy
import networkx as nx
import json
from networkx.readwrite import json_graph


def convert_json_to_npz():
    # load node name id mapping
    _name_id_map = json.load(open("reddit" + "-id_map.json"))
    num_node_map = len(_name_id_map)
    # convert node_id to be int
    if isinstance(list(_name_id_map.values())[0], int):
        convert_nodeid_to_int = lambda n: n
    else:
        convert_nodeid_to_int = lambda n: int(n)
    #name_id_map = dict([[node_name, convert_nodeid_to_int(node_id)] for node_name, node_id in _name_id_map.items()])
    #print("name_id_map", len(name_id_map), name_id_map)
    id_name_map = dict([[convert_nodeid_to_int(node_id), node_name] for node_name, node_id in _name_id_map.items()])
    # print("id_name_map", len(id_name_map), id_name_map)
    _class_map = json.load(open("reddit" + "-class_map.json"))
    _feats = np.load("reddit" + "-feats.npy")
    num_node_fea = _feats.shape[0]
    num_node_class = len(_class_map)
    print("load class and feature finished!")

    G_data = json.load(open("reddit" + "-G_full.json"))
    G = json_graph.node_link_graph(G_data)
    print("node_link_graph created!")
    num_node_G = G.number_of_nodes()
    # sanity check
    assert num_node_fea == num_node_class == num_node_G == num_node_map
    num_of_edge = G.number_of_edges()
    num_of_node = num_node_G
    print("num of nodes", num_of_node)
    print("num of edges", num_of_edge)

    node_ids = np.arange(0, num_of_node, dtype=np.int32)
    features = _feats[node_ids]

    node_name_type_G = lambda n : n
    node_name_list_from_mapping = [node_name_type_G(id_name_map[node_id]) for node_id in node_ids]

    adj = nx.to_scipy_sparse_matrix(G=G, nodelist=node_name_list_from_mapping, format='coo')
    scipy.sparse.save_npz("reddit_graph.npz", adj)
    # Types of the nodes, 1 --> Train, 2 --> Valid, 3 --> Test
    node_types = np.array([int(G.node[node_name_type_G(id_name_map[node_id])]['test']) * 2 + \
                           int(G.node[node_name_type_G(id_name_map[node_id])]['val']) * 1 + 1 for node_id in node_ids],
                          dtype=np.int32)
    print("node_types", node_types.shape, node_types)

    labels = [_class_map[id_name_map[node_id]] for node_id in node_ids]
    labels = np.array(labels, dtype=np.int32)
    # feature scaling
    from sklearn.preprocessing import StandardScaler
    scaler = StandardScaler()
    train_indices = node_ids[node_types == 1]
    scaler.fit(features[train_indices, :])
    features = scaler.transform(features)
    print("features", features.shape)
    print("labels", labels.shape)

    np.savez_compressed('reddit_data.npz',
                        feature=features,
                        label=labels,
                        node_types=node_types,
                        node_ids=node_ids)


convert_json_to_npz()
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by ZiyueHuang** [comment]
@zheng-da

File: python/dgl/data/reddit.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+class RedditDataset(object):
+    def __init__(self):
+        download_dir = get_download_dir()
+        # TODO upload the .npz files to S3
+        # graph
+        coo_adj = sp.load_npz(os.path.join(download_dir, "reddit_graph.npz"))
+        self.graph = dgl.DGLGraph(coo_adj, readonly=True)
+        # features and labels
+        reddit_data = np.load(os.path.join(download_dir, "reddit_data.npz"))
```
Comment (zheng-da) [inline python/dgl/data/reddit.py]: do you reorganize your data? it's different from the ones I get from your script.

**Comment by ZiyueHuang** [comment]
add self loop reddit

```
    print(len(coo_adj.row), " num edges")
    print(coo_adj.shape, " adj shape")
    row = list(coo_adj.row)
    col = list(coo_adj.col)
    diag = coo_adj.diagonal()
    print(np.nonzero(diag), " empty diag")
    n = len(diag)
    row.extend([i for i in range(n)])
    col.extend([i for i in range(n)])
    nn = len(row)
    self_loop_coo = sp.coo_matrix((np.ones((nn,)), (np.array(row), np.array(col))), shape=(n, n))
    print(self_loop_coo.shape, " self_loop_adj shape")
    sp.save_npz("/home/ubuntu/reddit/reddit_self_loop_graph.npz", self_loop_coo)
```

File: python/dgl/data/reddit.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import scipy.sparse as sp
+import numpy as np
+import dgl
+import os, sys
+from ..graph_index import create_graph_index
+from .utils import download, extract_archive, get_download_dir, _get_dgl_url
+
+
+class RedditDataset(object):
+    def __init__(self, self_loop=False):
```
Comment (zheng-da) [inline python/dgl/data/reddit.py]: how do we use self_loop argument?

**Comment by ZiyueHuang** [comment]
`--dataset reddit`

`--dataset reddit-self-loop`

**Pull request merged**